main/ipset/ipset.initd

```diff
@@ -59,57 +98,141 @@ stop() {
 save() {
 	ebegin "Saving firewall IP sets"
 
-	ipset save | while read cmd; do
-		set -- $cmd
-		local action=$1
-		local file=$DIR/$2
-		shift 2
-		if [ "$action" = create ]; then
-			echo $* > $file
-		elif [ "$action" = add ]; then
-			echo $* >> $file
-		fi
-	done
-
-	for name in $(set_files); do
-		set_exists $name || rm -f $DIR/$name
-	done
+	einfo "--> ${IPSET_SAVE}"
+	$IPSET save > ${IPSET_SAVE}
 
 	eend $STATUS
 }
 
 reload() {
-	ebegin "Loading firewall IP sets"
-
 	local swap=
+	ebegin "Reloading all firewall IP sets"
+
 	for name in $(set_files); do
 		local new=$name
 		if set_exists $name; then
 			new=_init_$name
 			swap="$swap $name"
+			create_setfile $new $DIR/$name
 		fi
-		ipset create $new $(set_file $name | head -n 1)
-	done
-
-	for name in $(set_files); do
-		local new=$name
-		set_exists _init_$name && new=_init_$name
-		set_file $name | tail -n +2 | while read m; do
-			ipset add $new $m
-		done
 	done
 
 	for name in $swap; do
-		ipset swap $name _init_$name
+		einfo "swapping _init_$name --> $name"
+		$IPSET swap _init_$name $name
+		ewarn "destroying ipset: _init_$name"
+		$IPSET destroy _init_$name
 	done
 
-	for name in $(sets list:set); do
-		[ -f $DIR/$name ] || ipset destroy $name
+	eend $STATUS
+	status
+}
+
+load() {
+	local name= ldsets=$(sets) tmpfile=$(mktemp)
+	ebegin "Loading inactive IP sets in: $DIR"
+
+	for name in $(set_files); do
+		if ! echo $ldsets |grep -q $name; then
+			create_setfile $name $DIR/$name
+			if set_exists $name; then
+				add_ip $name
+			fi
+		fi
 	done
+}
 
-	for name in $(sets); do
-		[ -f $DIR/$name ] || ipset destroy $name
+create_setfile() {
+	local tmpfile=$(mktemp) name=$1 file=$2
+	local  set_format=$(head -n 1 $file)
+	einfo "creating ipset: '$name' => type: $set_format"
+
+	# remove any comments
+	set_file $file > $tmpfile
+
+	# loading from a file is much faster
+	# than adding elements individually
+	sed -i "s/^/add $name /" $tmpfile
+	sed -i "1s/.*/create $name $set_format/" $tmpfile
+
+	$IPSET restore -exist -f $tmpfile
+	rm -f $tmpfile
+}
+unload() {
+	local name= ldsets=$(sets) i=0 choice=
+	if [ -n "$ldsets" ]; then
+		for name in $ldsets; do
+			i=$(( $i + 1 ))
+			einfo "($i): $name"
+		done
+		printf "\nSelect ipset to destroy? [1 - $i]: "; read choice
```

Comment:
From a cursory search, this is the only init script in aports that has an interactive terminal read with no timeout. The nearest other script is cryptsetup/dmcrypt.initd .